dev/buildtool/git.py

```diff
@@ -0,0 +1,762 @@
+      if at_index == self.MINOR_INDEX:
+        minor += 1
+      elif at_index == self.MAJOR_INDEX:
+        minor = 0
+        major += 1
+      else:
+        raise ValueError('Invalid index={0}'.format(at_index))
+
+    return SemanticVersion(self.series_name, major, minor, patch)
+
+
+class CommitTag(
+    collections.namedtuple('CommitTag', ['commit_id', 'tag', 'version'])):
+  """Denotes an individual result of git show-ref --tags."""
+
+  @staticmethod
+  def make(ref_line):
+    """Create a new instance from a response line.
+
+    Args:
+      ref_line: [string] Response from "git show-ref --tags"
+    """
+    # ref_line is in the form "$commit_id refs/tags/$tag"
+    tokens = ref_line.split(' ')
+    line_id = tokens[0]
+    tag_parts = tokens[1].split('/')
+    tag = tag_parts[len(tag_parts) - 1]
+    version = LooseVersion(tag)
+    return CommitTag(line_id, tag, version)
+
+  @staticmethod
+  def compare_tags(first, second):
+    """Comparator for instances compares the lexical order of the tags."""
+    return cmp(first.tag, second.tag)
+
+  # pylint: disable=multiple-statements
+  def __lt__(self, other): return self.tag.__lt__(other.tag)
+  def __le__(self, other): return self.tag.__le__(other.tag)
+  def __eq__(self, other): return self.tag.__eq__(other.tag)
+  def __ge__(self, other): return self.tag.__ge__(other.tag)
+  def __gt__(self, other): return self.tag.__gt__(other.tag)
+  def __ne__(self, other): return self.tag.__ne__(other.tag)
+
+
+class CommitMessage(
+    collections.namedtuple('CommitMessage',
+                           ['commit_id', 'author', 'date', 'message'])):
+  """Denotes an individual entry in 'git log --pretty'."""
+  _MEDIUM_PRETTY_COMMIT_MATCHER = re.compile(
+      '(.+)\nAuthor: *(.+)\nDate: *(.*)\n', re.MULTILINE)
+
+  _EMBEDDED_COMMIT_MATCHER = re.compile(
+      r'^( *)commit [a-f0-9]+\n'
+      r'^\s*Author: .+\n'
+      r'^\s*Date:   .+\n',
+      re.MULTILINE)
+
+  _EMBEDDED_SUMMARY_MATCHER = re.compile(
+      r'^\s*(?:\*\s*)?[a-z]+\(.+?\): .+',
+      re.MULTILINE)
+
+  # The vocabulary in the following list was taken looking at what's
+  # used in practice (right or wrong), for conforming log entries,
+  # using the following command over the various spinnaker repositories:
+  #
+  # git log --pretty=oneline \
+  #   | sed "s/^[a-f0-9]\+ //g" \
+  #   | egrep "^[^ \(]+\(" \
+  #   | sed "s/^\([^\(]\+\).*/\1/g" \
+  #   | sort | uniq
+  DEFAULT_PATCH_REGEXS = [
+      # Some tags indicate only a patch release.
+      re.compile(r'^\s*'
+                 r'(?:\*\s+)?'
+                 r'((?:fix|bug|docs?|test)[\(:].*)',
+                 re.MULTILINE)
+  ]
+  DEFAULT_MINOR_REGEXS = [
+      # Some tags indicate a minor release.
+      # These are features as well as hints to non-trivial
+      # implementation changes that suggest a higher level of risk.
+      re.compile(r'^\s*'
+                 r'(?:\*\s+)?'
+                 r'((?:feat|feature|chore|refactor|perf|config)[\(:].*)',
```

Comment:
We can do that. The old rules werent explicit (or at least not clear). It looked like it was looking for features and breaking changes but that was it. I'm trying to be explicit here and guessed at what I thought the rules were. The intent wasnt to change the rules. I assumed config changes were like feature changes. Major versions to me are things that break compatibility across components as opposed to within the component. If I'm wrong we can change it now, or talk about it later and change it later. We can change it again later either way. It isnt going online for a while since I havent refactored the post-validation steps yet.

Reply:
Let's not worry about it in this PR and revisit `config` as a breaking change later as a group.

Reply:
I'm going to submit like this for now, but as noted above and offline, I'm completely open to changing it in a future PR.